---
title: "Breastcancer Gene Expression Study: KPNA2 gene"
author: "Lieven Clement"
output:
  html_notebook: default
---

#1. Background
Background: Histologic grade in breast cancer provides clinically important prognostic information. Researchers examined whether histologic grade was associated with gene expression profiles of breast cancers and whether such profiles could be used to improve histologic grading. In this tutorial we will assess the impact of histologic grade on expression of the KPNA2 gene that is known to be associated with poor BC prognosis.
The patients, however, do not only differ in the histologic grade, but also on their lymph node status. The lymph nodes were not affected (0) or surgically removed (1).


#2. Data analysis
##2.1. Import KPNA2 data in R
```{r}
kpna2 <- read.table("gse2990BreastcancerOneGene.txt",header=TRUE)
head(kpna2)
```

##2.2. Transform the variables 

Transform grade and node to a factor
```{r}
kpna2$grade <- as.factor(kpna2$grade)
kpna2$node <- as.factor(kpna2$node)
```

Transform response to a log2-scale

```{r}
kpna2$logExp=log2(kpna2$gene)
```

##2.3. Data exploration
Histological grade and lymph node status can both have an effect on the KPNA2 gene. Moreover, it is also possible that the effect of the histological grade changes according to the lymph node status. Therefore, we plot the gene expression for each grade x node combination. 

```{r}
library(ggplot2)
plot1 <- ggplot(kpna2, aes(x=node:grade, y=logExp,fill=grade:node))
plot2 <- ggplot(kpna2, aes(x=age, y=logExp,colour=grade:node))
plot3 <- ggplot(kpna2, aes(x=size, y=logExp,colour=grade:node))
plot1 +geom_boxplot(outlier.shape=NA) + geom_point(position=position_jitter(width=.1))
plot2 + geom_point()
plot3 + geom_point()
```
 
 

The plot suggests

- an effect of the histological grade
- an effect of node status
- The differential expression associated to grade seems to differ according to the lymph node status (interaction) 

##2.4. Hypotheses of interest

The researchers focus on the effect of the hystological grade, but because we expect a different effect in patients with irradicated nodes and unaffected nodes we will have to study the effect of hystological grade within patients with a certain node status. 

Research questions:

- Is KPNA2 expression on average different between grade 3 and grade 1 in patients with unaffected lymph nodes
- Is KPNA2 expression on average different between grade 3 and grade 1 in patients with affected lymph nodes that were surgically removed
- Is the fold change of KPNA2 expression between grade 3 and grade 1 patients on average different according to the lymph node status? (interaction)

##2.5. Model
Histologic grade and lymph node status can have an effect on the kpna2 gene. Moreover, it is also possible that the differential expression due to histological grade is differt in patients that have unaffected lymph nodes and patients for which the lymph nodes had to be removed. Hence, we will have to model the gene expression by using main effects for grade, node and a grade x node interaction. 

```{r}
#Adapt model!
fit <- lm(logExp~grade*node,data=kpna2)
plot(fit)
```


The variance is more or less equal for every treatment x node combination. 
The QQ-plot of the residuals shows no severe deviations from normality.

```{r}
summary(fit)
#Calculate confidence intervals for parameters of model
confint(fit)

#Transform parameters and the CI back to the original scale
2^fit$coef
2^confint(fit)
```



##2.6. Interpretation of model parameters and statistical tests
We model the log$_2$-transformed intensities with the following model: 
$$
y=\beta_0+\beta_{g3}x_{g3}+\beta_{n1}x_{n1}+\beta_{g3n1}x_{g3}x_{n1},
$$

with $\beta_0$ the intercept, $\beta_{g3}$ the main effect for grade, $x_{g3}$ a dummy variable for grade which is 0 for the control treatment in the absence of grade and 1 for the treatment with grade, $\beta_{n1}$ the main effect for node, $x_{n1}$ a dummy variable that is 0 for the measurements of patients with unaffected lymph nodes and 1 for patients for which the lymph nodes were removed and $\beta_{g3n1}$ the interaction effect between grade and node. 
To ease the interpertation of the parameters, $\log_2$ transformed mean intensities are given for each treatment group as well as corresponding contrasts between treatments, which have an interpretation in terms of $\log_2$ transformed fold changes (FC). 


- $\log_2\hat{\mu}_{g1n0}=\hat\beta_0$, $\log_2 \hat{\mu}_{g3n0}=\hat\beta_0+\hat\beta_{g3}$ --> $\log_2 \widehat{FC}_{g3n0-g1n0}=\hat\beta_{g3}$

- $\log_2 \hat {\mu}_{g1n1}=\hat\beta_0+\hat\beta_t$, $\log_2 \hat{\mu}_{g3n1}=\hat\beta_0+\hat\beta_{g3}+\hat\beta_t+\hat\beta_{g3n1}$ --> $\log_2 \widehat{FC}_{g3n1-g1n1}=\hat \beta_{g3} +\hat\beta_{g3n1}$

- $\log_2 \widehat{FC}_{g1n1-g1n0}=\hat\beta_{n1}$, $\log_2 \widehat{FC}_{g3n1-g3n0}=\hat\beta_{n1}+\hat\beta_{g3n1}$ --> $\log_2\frac{\widehat{FC}_{g3n1-g1n1}}{\widehat{FC}_{g3n0-g1n0}}=\hat\beta_{g3n1}$

with $\log_2\hat \mu_{g1n0}$, $\log_2\hat \mu_{g3n0}$, $\log_2\hat \mu_{g1n1}$ and $\log_2\hat \mu_{g3n1}$ the estimated mean $\log_2$ transformed intensity for patients with grade 1 and node 0 status, grade 3 and node 0 status, grade 1 and node 1 status and grade 3 and node 1 status, respectively. With $\log_2 \widehat{FC}_{b-a}$ we indicate $\log_2$ transformed fold change estimates between treatment b and treatment a, i.e. $\log_2 \widehat{FC}_{b-a}=\log_2 \hat\mu_{b}-\log_2 \hat\mu_a=\log_2 \frac{\hat\mu_{b}}{\hat\mu_{a}}$.

The model immediately provides statistical tests for assessing the significance of fold changes between grade 3 and grade 1 for patients with unaffected lymph nodes (n=0) $\log_2 \widehat{FC}_{g3n0-g1n0}$,  fold changes between the grade 1-node 1 patients and grade 1- node 0 patients $\log_2 \widehat{FC}_{g1n1-g3n0}$ and for differences in fold change due to histological grade for node 1 patients and node 0 patients. $\log_2\frac{\widehat{FC}_{g3n1-g1n1}}{\widehat{FC}_{g3n0-g1n0}}$, the interaction term. 
```{r}
summary(fit)
```

And confidence intervals
```{r}
CIfit<-confint(fit)
CIfit
```

We can backtransform the parameters and confidence intervals to get estimates and CI for the Fold changes on the original scale

```{r}
2^coefficients(fit)
2^CIfit
```

- The average intensity for grade 1 patients with unaffected lymph nodes equals $\exp(\hat \beta_0)$=
`r round(2^fit$coef["(Intercept)"],2)`.
- The hystological grade effect in patients with unaffected lymph nodes and the lymph node effect in grade 1 patients are extremely significant (p<<0.001) and very significant (p=0.002), respectively. 
	- When lymph nodes are unaffected, the expression is on average `r round(2^fit$coef["grade3"],2)` times higher for patients with histological grade 3 than patients with histological grade 1 (95% CI [`r round(2^CIfit["grade3",],2)`]). 
	- The gene expression in histological grade 1 patients is on average `r round(1/2^fit$coef["node1"],2)` times higher when the lymph nodes are affected (95% CI [`r round(1/2^CIfit["node1",],2)`]).
- The difference in histological grade effect between patients with and without affected lymph nodes is very significant (p=0.007). The fold change due to histological grade is on average `r round(2^fit$coef["grade3:node1"],2)` times lower in patients with affected lymph nodes than in patients with unaffected lymph nodes (95% CI [`r round(2^CIfit["grade3:node1",],2)`]).


##2.7. Histological grade effect in patients with affected lymph nodes. 

The researchers, however, are also interested in testing the significance of the fold change between patients with grade 3 and grade 1 tumors that had their lymph nodes removed. This involves assessing a contrast of the model parameters: $\log_2 \widehat{FC}_{g3n1-g1n1}=\hat \beta_{g3} +\hat\beta_{g3n1}$.

###2.7.1. Test for contrast

With the multcomp package you can define contrasts, linear combinations of model parameters and assess if the contrast is statistically significant. 

$$H_0: \mathbf{L}^T\boldsymbol{\beta}=0$$
vs 
$$H_1: \mathbf{L}^T\boldsymbol{\beta}\neq 0,$$

with 
$$
L=\left[\begin{matrix}0\\1\\0\\1\end{matrix}\right] \text{and } \boldsymbol{\beta}=\left[\begin{matrix}\beta_0\\\beta_{g3}\\\beta_{n1}\\\beta_{g3n1}\end{matrix}\right]
$$
for the contrast of interest. 

In R, we can either use a vector L or define a formula using the parameter names. We first illustrate the latter.

```{r}
library(multcomp)
contrast <- glht(fit, linfct = c("grade3 +grade3:node1 = 0"))
summary(contrast)
confint(contrast)
```

We can also define the vector ourself.

```{r}
L<-matrix(nrow=length(coef(fit)),ncol=1,0)
rownames(L) <-names(coef(fit))
colnames(L) <-"contrast1"
L[grepl(rownames(L),pattern="grade3"),1] <- 1
L
```
and assess the significance of the contrast by 
```{r}
contrastVec <- glht(fit, linfct = t(L))
summary(contrastVec)
```

###2.7.2. Confidence intervals FC for effect of grade in patients with removed lymph nodes.
```{r}
CIFCg3n1_g1n1 <- confint(contrastVec)
CIFCg3n1_g1n1
```

Backtransform to FC
```{r}
2^CIFCg3n1_g1n1$confint
```

The histological grade effect in patients with affected lymph nodes is also extremely significant (p<<0.001).
The expression of the kpna2 gene in patients with affected lymph nodes is on average `r  round(2^CIFCg3n1_g1n1$confint,2)[1]` times higher in patients with a histological grade 3 tumor than in patients with histological grade 1 tumor. (95% CI [`r round(2^CIFCg3n1_g1n1$confint,2)[-1]`]).

##2.8. Assess all hypotheses of interest and correct for multiple testing
You can also define multiple contrasts and assess them simultaneously, then the multcomp package will automatically adjust for multiple testing.

```{r}
L<-matrix(nrow=length(coef(fit)),ncol=3,0)
rownames(L) <-names(coef(fit))
colnames(L) <-paste0("contrast",1:3)
L["grade3",1]<-1
L[grepl(rownames(L),pattern="grade3"),2] <- 1
L["grade3:node1",3]<-1
L
```

Complete the analysis for all contrasts!

#3. Conclusions

#4. Redo the analysis by coding the linear regression analysis yourself using matrices. 

##4.1. Fit

##4.2. Standard errors

##4.3. Inference with contrasts

